src/Raven.Server/Documents/AI/Embeddings/EmbeddingsGenerator.cs

```diff
@@ -0,0 +1,908 @@
+        {
+            try
+            {
+                IList<ReadOnlyMemory<float>> allEmbeddings;
+
+                try
+                {
+                    allEmbeddings = await AiHelper.GenerateEmbeddingsAsync(_embeddingGenerationService, batch, _cancellationToken);
+                }
+                catch (HttpOperationException httpOperationException) when (httpOperationException.StatusCode == HttpStatusCode.TooManyRequests)
+                {
+                    throw new EmbeddingGenerationException(
+                        $"Failed to generate embeddings due to rate limits. Consider decreasing the number of elements processed in a single batch " +
+                        $"('{RavenConfiguration.GetKey(x => x.Ai.EmbeddingsGenerationMaxBatchSize)}') or increasing the " +
+                        $"limits on your model deployment.", httpOperationException);
+                }
+
+
+                PortableExceptions.ThrowIf<IOException>(allEmbeddings.Count != batch.Count, "Model returned a different count of embeddings than expected");
+
+                for (int i = 0; i < allEmbeddings.Count; i++)
+                {
+                    embeddings[i].Add(EmbeddingsHelper.CreateEmbeddingValue(allEmbeddings[i], Configuration.Quantization));
+                }
+
+                if (_parent._mode is Mode.Query)
+                {
+                    // for queries we want to release the thread ASAP and send it the embeddings
+                    foreach (var work in works)
+                    {
+                        work.TaskCompletionSource.TrySetResult();
+                        // explicitly *not* calling this here, we have a value
+                        // but we haven't persisted that yet
+                        // work.Owner.RemoveFromCache(ge.CacheKey);
+                    }
+                }
+                _parent.ProcessInBackground(new StoreEmbeddings(works, Configuration.Quantization));
+            }
+            catch (Exception e)
+            {
+                foreach (var work in works)
+                {
+                    work.TaskCompletionSource.TrySetException(e);
+                    work.Owner.RemoveFromCache(work.CacheKey);
+                }
+            }
+        }
+
+        public bool ModifiedFrom(EmbeddingsGenerationConfiguration updated, AiConnectionString updateConnectionString)
+        {
+            return Configuration.Compare(updated) != EtlConfigurationCompareDifferences.None ||
+                   _connectionString.Compare(updateConnectionString) != AiSettingsCompareDifferences.None;
+        }
+
+        public void Wake()
+        {
+            _hasWork.Set();
+        }
+    }
+    
+    private void ProcessInBackground(object works)
+    {
+        _toCache.Enqueue(works);
+        _hasWork.Set();
+    }
+    private AiWorker CreateAiWorker(EmbeddingsGenerationTaskIdentifier id)
+    {
+        var record = _database.ReadDatabaseRecord();
```

Comment:
better to use ReadRawDatabaseRecord (and dispose it early) ?

Reply:
That isn't available here (only at the cluster level).
I don't care about perf here, this is meant to be called _once_ typically for the lifetime of a db